### PR TITLE
Dvnagorny/sklearn

### PR DESCRIPTION
**Pull request by dvnagorny**

### CODE REVIEW

**Review by anton-malakhov — APPROVED** [review]
Good for starters

File: papers/oleksandr_pavlyk/optimizations.rst [hunk 1]
```diff
@@ -72,9 +72,129 @@ The most significant improvements in memory optimizations in Update 2 comes from
 
 Faster Machine Learning with Scikit-learn
 -----------------------------------------
-Scikit-learn is among the most popular Python machine learning packages. The initial release of Intel Distribution for Python provided Scikit-learn optimizations via respective NumPy and SciPy functions accelerated by Intel MKL. Update 2 optimizes selective key machine learning algorithms in Scikit-learn, accelerating them with the Intel |R| Data Analytics Acceleration Library (Intel |R| DAAL).
+Scikit-learn is well know library that provides a log of algorithms for many areas of machine learning.
```
Comment (oleksandr-pavlyk) [inline papers/oleksandr_pavlyk/optimizations.rst]: It should be "... that provides a **lot** of algorithms ..."

File: papers/oleksandr_pavlyk/optimizations.rst [hunk 2]
```diff
@@ -72,9 +72,129 @@ The most significant improvements in memory optimizations in Update 2 comes from
...
+It is originally written in C++ and provides Java and Python bindings.
+In spite of the fact that DAAL is heavily optimized for all Intel Architectures including Xeon Phi it’s very questionable how to use DAAL binding from Python.
+DAAL bindings for python are generated automatically and reflects original C++ API very close that makes its usage quite complicated because of not native for python programmers idioms and chary documentation.
+
+To mix the power of well optimized for HW native code with familiar ML API Intel Python Distribuition started efforts on scikit-learn optimization.
+Therefore, beginning from 2017 U2 IDP includes scikit-learn with daal4sklearn module.
 Specifically, Update 2 optimizes Principal Component Analysis (PCA), Linear and Ridge Regressions, Correlation and Cosine Distances, and K-Means. Speedups may range from 1.5x to 160x.
 
+There are no exact matching between sklearn and DAAL API and they aren’t fully compatible for all inputs so for the cases when daal4sklearn detects incompatibility it fallbacks to original sklearn’s implementation.
```
Comment (oleksandr-pavlyk) [inline papers/oleksandr_pavlyk/optimizations.rst]: It should be "There is no exact matching ..."

Comment (oleksandr-pavlyk) [reply papers/oleksandr_pavlyk/optimizations.rst]: It should " ... it falls back on ...", see [TFD:fall back on](http://www.tfd.com/fall+back+on).

File: papers/oleksandr_pavlyk/optimizations.rst [hunk 3]
```diff
@@ -72,9 +72,129 @@ The most significant improvements in memory optimizations in Update 2 comes from
...
+
+Daal4sklearn is enabled by default but provides simple API that allows disabling its functionality:
```
Comment (oleksandr-pavlyk) [inline papers/oleksandr_pavlyk/optimizations.rst]: "... bug provides **a** simple API ..."

Comment (oleksandr-pavlyk) [reply papers/oleksandr_pavlyk/optimizations.rst]: "... that allows **to disable** its functionality:"

File: papers/oleksandr_pavlyk/optimizations.rst [hunk 4]
```diff
@@ -72,9 +72,129 @@ The most significant improvements in memory optimizations in Update 2 comes from
...
+        def train(X):
+            kmeans.fit(X)
+
+        for p, n in problem_sizes:
+            print (p,n, end=' ')
+            X_local = X[(p,n)]
+            train(X_local)
+            print('')
+
+Using all 32 cores of Xeon E5-2698v3 IDP’s KMeans can be faster more than 50 times comparing with python available on Ubuntu 14.04.
```
Comment (oleksandr-pavlyk) [inline papers/oleksandr_pavlyk/optimizations.rst]: "... more than 50 times **compared** with system **Python** that comes with Ubuntu 14.04.

File: papers/oleksandr_pavlyk/optimizations.rst [hunk 5]
```diff
@@ -72,9 +72,129 @@ The most significant improvements in memory optimizations in Update 2 comes from
...
+ |10000 |50  |0,09     |0,02      |2,21        |0,87         |23,83        |40,76         |
+ |50000 |2   |0,08     |0,01      |1,62        |0,57         |20,57        |47,43         |
+ |50000 |25  |0,67     |0,07      |14,43       |2,79         |21,47        |38,69         |
+ |50000 |50  |1,05     |0,10      |24,04       |4,00         |22,89        |38,52         |
+ |100000|2   |0,15     |0,02      |3,33        |0,87         |22,30        |56,72         |
+ |100000|25  |1,34     |0,11      |33,27       |5,53         |24,75        |49,07         |
+ |100000|50  |2,21     |0,17      |63,30       |8,36         |28,65        |47,95         |
+ +------+----+---------+----------+------------+-------------+-------------+--------------+
+
+We compared the similar runs for other algorithms and normalized results by results obtained with DAAL in C++ without python to estimate overhead from python wrapping.
```
Comment (oleksandr-pavlyk) [inline papers/oleksandr_pavlyk/optimizations.rst]: "We performed similar runs for other algorithms, reporting execution times as a fraction of execution time of C++ native program that uses DAAL in order to estimate the overhead of using Python wrappers."

**Comment by anton-malakhov** [comment]
@dvnagorny please move your .rst file and all the corresponding attachments into a subfolder like Andres did

File: papers/oleksandr_pavlyk/sklearn.rst [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        from sklearn.cluster import KMeans
+
+        import argparse
+        argParser = argparse.ArgumentParser(prog="pairwise_distances.py",
+                                            description="sklearn pairwise_distances benchmark",
+                                            formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+
+        argParser.add_argument('-i', '--iteration', default='10', help="iteration", type=int)
+        argParser.add_argument('-p', '--proc', default=-1, help="n_jobs for algorithm", type=int)
+        args = argParser.parse_args()
```
Comment (anton-malakhov) [inline papers/oleksandr_pavlyk/sklearn.rst:43]: I don't think we need all that code here. Let's publish it on github and provide the link here. Please  leave only vital part of code snippets @dvnagorny

**Pull request merged**